core/src/main/java/org/elasticsearch/search/sort/FieldSortBuilder.java

```diff
@@ -220,6 +234,56 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
     }
 
     @Override
+    public SortField build(QueryShardContext context) throws IOException {
+        if (DOC_FIELD_NAME.equals(fieldName)) {
+            if (order == SortOrder.DESC) {
+                return SORT_DOC_REVERSE;
+            } else {
+                return SORT_DOC;
+            }
+        } else {
+            MappedFieldType fieldType = context.fieldMapper(fieldName);
+            if (fieldType == null) {
+                if (unmappedType != null) {
+                    fieldType = context.getMapperService().unmappedFieldType(unmappedType);
+                } else {
+                    throw new QueryShardException(context, "No mapping found for [" + fieldName + "] in order to sort on");
+                }
+            }
+
+            if (!fieldType.isSortable()) {
+                throw new QueryShardException(context, "Sorting not supported for field[" + fieldName + "]");
+            }
+
+            // Enable when we also know how to detect fields that do tokenize, but only emit one token
+            /*if (fieldMapper instanceof StringFieldMapper) {
+                StringFieldMapper stringFieldMapper = (StringFieldMapper) fieldMapper;
+                if (stringFieldMapper.fieldType().tokenized()) {
+                    // Fail early
+                    throw new SearchParseException(context, "Can't sort on tokenized string field[" + fieldName + "]");
+                }
+            }*/
+
+            MultiValueMode localSortMode = null;
+            if (sortMode != null) {
+                localSortMode = MultiValueMode.fromString(sortMode.toString());
+            }
+            if (fieldType.isNumeric() == false && (sortMode == SortMode.SUM || sortMode == SortMode.AVG)) {
+                throw new QueryShardException(context, "we only support AVG and SUM on number based fields");
+            }
+            boolean reverse = (order == SortOrder.DESC);
+            if (localSortMode == null) {
+                localSortMode = reverse ? MultiValueMode.MAX : MultiValueMode.MIN;
```

Comment:
Hmpf - should first read through and comment afterwards. I saw only now that we need "reverse" further below anyway :/